src/openms/source/ANALYSIS/ID/PrecursorPurity.cpp

```diff
@@ -0,0 +1,231 @@
+
+namespace OpenMS
+{
+
+  PrecursorPurity::PurityScores PrecursorPurity::computePrecursorPurity(const PeakSpectrum& ms1, const Precursor& pre, const double precursor_mass_tolerance, const bool precursor_mass_tolerance_unit_ppm)
+  {
+    PrecursorPurity::PurityScores score;
+    double target_mz = pre.getMZ();
+    double lower = target_mz - pre.getIsolationWindowLowerOffset();
+    double upper = target_mz + pre.getIsolationWindowUpperOffset();
+    int charge = pre.getCharge();
+
+    double precursor_tolerance_abs = precursor_mass_tolerance_unit_ppm ? (target_mz * precursor_mass_tolerance*2 * 1e-6) : precursor_mass_tolerance*2;
+
+    auto lower_it = ms1.MZBegin(lower);
+    auto upper_it = ms1.MZEnd(upper);
+
+    // std::cout << "charge: " << charge << " | lower: " << lower << " | target: " << target_mz << " | upper: " << upper << std::endl;
+    // std::cout << "lower offset: " << pre.getIsolationWindowLowerOffset() << " | upper offset: " << pre.getIsolationWindowUpperOffset() << std::endl;
+    // std::cout << "lower peak: " << (*lower_it).getMZ() << " | upper peak: " << (*upper_it).getMZ() << std::endl;
+
+    PeakSpectrum isolated_window;
+    while (lower_it != upper_it)
+    {
+      isolated_window.push_back(*lower_it);
+      lower_it++;
+    }
+
+    // std::cout << "Isolation window peaks: " << isolated_window.size();
+    for (auto peak : isolated_window)
+    {
+      // std::cout << " | " << peak.getMZ();
+    }
+    // std::cout << std::endl;
+
+    // total intensity in isolation window
+    double total_intensity(0);
+    double target_intensity(0);
+    Size target_peak_count(0);
+    for (auto peak : isolated_window)
+    {
+      total_intensity += peak.getIntensity();
+    }
+
+    // search for the target peak, return scores with 0-values if it is not found
+    if (isolated_window.empty())
+    {
+      return score;
+    }
+
+    // estimate a lower boundary for isotopic peaks
+    int negative_isotopes((pre.getIsolationWindowLowerOffset() * charge));
+    double iso = -negative_isotopes;
+    // depending on the isolation window, the first estimated peak might be outside the window
+    if (target_mz + (iso * Constants::C13C12_MASSDIFF_U / charge) < lower)
+    {
+      iso++;
+    }
+
+    // std::cout << "target peaks: ";
+    // deisotoping (try to find isotopic peaks of the precursor mass, even if the actual precursor peak is missing)
+    while (true) // runs as long as the next mz is within the isolation window
+    {
+      double next_peak = target_mz + (iso * Constants::C13C12_MASSDIFF_U / charge);
+      // std::cout << iso << " : iso | " << next_peak << " : next peak | ";
+
+      // stop loop when new mz is outside the isolation window
+      // changes through the isotope index iso
+      if (next_peak > upper)
+      {
+        break;
+      }
+      int next_iso_index = isolated_window.findNearest(next_peak, precursor_tolerance_abs);
+      if (next_iso_index != -1)
+      {
```

Comment:
it seems here you are only using the closest peak, this will not work well for profile data or poorly centroided data.

Why not simply sum up all intensity in `next_peak +/- tolerance` ?

Reply:
yes, that is a problem I have already run into. For centroided data it seems to work better after running `HighResPrecursorMassCorrector` with the `nearest_peak' method. Will summing up the intensity within the tolerance window not also catch some noise? I am also not sure what tolerance to use here in general, for now I use the precursor search tolerance from my search engine * 2, because the 10 ppm I usually use for the search engine is too narrow.

Reply:
> Will summing up the intensity within the tolerance window not also catch some noise?

isnt that the point? 

> I am also not sure what tolerance to use here in general,

that should be the *resolution* of the instrument, e.g. how wide the peak is. Your 10 ppm is for the mass accuracy and there is a big difference (peak apex vs peak width)

Reply:
I think it makes most sense for centroided data. Resolution also might vary with m/z so I am not sure how to integrate this information properly

Reply:
should work when its in ppm

Reply:
I thought the resolution is about distinguishing two peaks and not about the total peak width (there are probably multiple definitions) so I am not totally convinced this works well for profile data. The accuracy for precursor mass assignment might be different than the accuracy in the MS1 (that's probably why he needs the precursor mass correction step). It will work for centroided data though but probably lead to effects summing up peaks twice, overlapping windows etc. For now I would go with a centroided only solution until more experiments are done.

Reply:
@hroest this part of the algorithm sums up the intensities of the "target peaks", meaning the precursor ion targeted for fragmentation and possible additional isotopic peaks. I sum these up here and remove the peaks from this sliced out spectrum. Everything else, all the peaks I can not match to isotopes of the precursor mass here, I consider as noise.
@timosachsenberg Yes I wrote this code mainly for centroided data, that is why I only consider the one closest peak for each theoretical isotopic mass. If there are additional peaks within the tolerance window, they do not match the isotopic pattern and are considered as noise. I think making it work for profile data would require a very different approach.